bofire/surrogates/mixed_tanimoto_gp.py

```diff
@@ -0,0 +1,369 @@
+                    ard_num_dims=len(ord_dims),
+                    active_dims=ord_dims,
+                )
+            ) + ScaleKernel(
+                mol_kernel_factory(
+                    batch_shape=aug_batch_shape,  # type: ignore
+                    ard_num_dims=len(mol_dims),
+                    active_dims=mol_dims,
+                )
+            )
+
+            prod_kernel = ScaleKernel(
+                cont_kernel_factory(
+                    batch_shape=aug_batch_shape,  # type: ignore
+                    ard_num_dims=len(ord_dims),
+                    active_dims=ord_dims,
+                )
+            ) * ScaleKernel(
+                mol_kernel_factory(
+                    batch_shape=aug_batch_shape,  # type: ignore
+                    ard_num_dims=len(mol_dims),
+                    active_dims=mol_dims,
+                )
+            )
+
+            covar_module = sum_kernel + prod_kernel
+
+        else:
+            sum_kernel = (
+                ScaleKernel(
+                    cont_kernel_factory(
+                        batch_shape=aug_batch_shape,  # type: ignore
+                        ard_num_dims=len(ord_dims),
+                        active_dims=ord_dims,
+                    )
+                )
+                + ScaleKernel(
+                    mol_kernel_factory(
+                        batch_shape=aug_batch_shape,  # type: ignore
+                        ard_num_dims=len(mol_dims),
+                        active_dims=mol_dims,
+                    )
+                )
+                + ScaleKernel(
+                    CategoricalKernel(
+                        batch_shape=aug_batch_shape,
+                        ard_num_dims=len(cat_dims),  # type: ignore
+                        active_dims=cat_dims,
+                        lengthscale_constraint=GreaterThan(1e-06),
+                    )
+                )
+            )
+
+            prod_kernel = (
+                ScaleKernel(
+                    cont_kernel_factory(
+                        batch_shape=aug_batch_shape,  # type: ignore
+                        ard_num_dims=len(ord_dims),
+                        active_dims=ord_dims,
+                    )
+                )
+                * ScaleKernel(
+                    mol_kernel_factory(
+                        batch_shape=aug_batch_shape,  # type: ignore
+                        ard_num_dims=len(mol_dims),
+                        active_dims=mol_dims,
+                    )
+                )
+                * ScaleKernel(
+                    CategoricalKernel(
+                        batch_shape=aug_batch_shape,
+                        ard_num_dims=len(cat_dims),  # type: ignore
+                        active_dims=cat_dims,
+                        lengthscale_constraint=GreaterThan(1e-06),
+                    )
+                )
+            )
+            covar_module = sum_kernel + prod_kernel
```

Comment:
One general question to the setup of the kernels: What happens if two molecules are involved, so two features with molecular features? Should we then not have two Tanimoto kernels?

Reply:
Good point. To create separate molecular kernels for each molecular input is possible to do. However, I am not sure how the performance would change with one overall Tanimoto kernel vs multiple separate Tanimoto kernels for each molecular feature.

I have not come across a use-case with suitable data for this yet internally or in the literature. Do you know of one? My assumptions, based on how the description of how the kernel works from Gauche, it sounds like one Tanimoto kernel across molecules should be fine the time being.

Nonetheless, I will definitely keep an open mind about this and amend it when there is a evaluation done with an actual dataset with more than one molecule.